src/commcare_cloud/commands/terraform/terraform_migrate_state.py

```diff
@@ -0,0 +1,256 @@
+        intermediate_state = state
+        print("Applying the following changes:{}".format(
+            ''.join('\n  - {:0>4} {}'.format(migration.number, migration.slug)
+                    for migration in unapplied_migrations)
+        ))
+        print("which will result in the following moves being made:")
+        migration_plans = []
+        for migration in unapplied_migrations:
+            print('  [{:0>4} {}]'.format(migration.number, migration.slug))
+            migration_plan = make_migration_plan(intermediate_state, migration)
+            intermediate_state = migration_plan.end_state
+            for start_address, end_address in migration_plan.moves:
+                print('    {} => {}'.format(start_address, end_address))
+            migration_plans.append(migration_plan)
+        if ask("Do you want to apply this migration?"):
+            for migration_plan in migration_plans:
+                migration = migration_plan.migration
+                print('  [{:0>4} {}]'.format(migration.number, migration.slug))
+                for start_address, end_address in migration_plan.moves:
+                    print('    {} => {}'.format(start_address, end_address))
+                    commcare_cloud(args.env_name, 'terraform', 'state', 'mv', start_address, end_address)
+            remote_migration_state_manager.push(RemoteMigrationState(number=migration.number, slug=migration.slug))
+
+
+Migration = namedtuple('Migration', 'number slug get_new_resource_address')
+MigrationPlan = namedtuple('MigrationPlan', 'migration start_state end_state moves')
+
+
+class RemoteMigrationState(jsonobject.JsonObject):
+    checkpoint = jsonobject.StringProperty()
+    slug = jsonobject.StringProperty()
+
+
+class RemoteMigrationStateManager(object):
+    def __init__(self, terraform_config):
+        self.aws_profile = terraform_config.aws_profile
+        self.environment = terraform_config.environment
+        self.state_bucket = terraform_config.state_bucket
+
+    @property
+    def s3_uri(self):
+        return 's3://{state_bucket}/migration-state/{environment}.json'.format(
+            state_bucket=self.state_bucket,
+            environment=self.environment,
+        )
+
+    @property
+    def env_vars(self):
+        env_vars = deepcopy(os.environ)
+        env_vars.update({
+            'AWS_PROFILE': self.aws_profile
+        })
+        return env_vars
+
+    def fetch(self):
+        """
+        Fetch remote migration state from S3
+
+        Essentially:
+
+        AWS_PROFILE={aws_profile} aws s3 cp s3://{state_bucket}/migration-state/{environment}.json {tempfile}
+
+        wrapped as as a RemoteMigrationState object.
+        """
+
+        temp_filename = tempfile.mktemp()
+        try:
+            command = ['aws', 's3', 'cp', self.s3_uri, temp_filename]
+            print_command(command)
+            try:
+                subprocess.check_output(command, env=self.env_vars, stderr=subprocess.STDOUT)
```

Comment:
why not use `boto` here?

Reply:
Yeah I probably should have. It's not currently in the install list, and I just didn't want to drop my headspace to figure out to do that and look up `boto3`'s API. But you're right, I should switch to that, and it should be pretty easy, so will do.